pkg/services/secrets/kvstore/migrations/datasources_mig.go

```diff
@@ -0,0 +1,85 @@
+package migrations
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/secrets/kvstore"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const (
+	dataSourceSecretType = "datasource"
+)
+
+type DataSourceSecretMigrationService struct {
+	sqlStore           *sqlstore.SQLStore
+	dataSourcesService datasources.DataSourceService
+	secretsStore       kvstore.SecretsKVStore
+	features           featuremgmt.FeatureToggles
+	log                log.Logger
+	bus                bus.Bus
+}
+
+func ProvideDataSourceMigrationService(
+	cfg *setting.Cfg, dataSourcesService datasources.DataSourceService,
+	secretsStore kvstore.SecretsKVStore, features featuremgmt.FeatureToggles,
+	sqlStore *sqlstore.SQLStore, bus bus.Bus,
+) *DataSourceSecretMigrationService {
+	return &DataSourceSecretMigrationService{
+		sqlStore:           sqlStore,
+		dataSourcesService: dataSourcesService,
+		secretsStore:       secretsStore,
+		features:           features,
+		log:                logger,
+		bus:                bus,
+	}
+}
+
+func (s *DataSourceSecretMigrationService) Run(ctx context.Context) error {
+	return s.sqlStore.InTransaction(ctx, func(ctx context.Context) error {
```

Comment:
Sounds good! Will update.

Reply:
Done on https://github.com/grafana/grafana/pull/50463/commits/520ee5e6d6359a9e8a1fd0ae8ef635cc9a8ea174

Reply:
InTransaction ended up completely removed since it's already called inside the UpdateDataSource function and since that same function Set and clear the secret based on the compatibility flag, it's not needed to call other functions that execute SQL operations.